modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+      if headBlock.isEmpty {
+        var tmp = headBlock;
+        headBlock = headBlock.next;
+        delete tmp;
+
+        if headBlock == nil then tailBlock = nil;
+      }
+    }
+
+    return arr;
+  }
+
+  inline proc takeElement() {
+    if isEmpty {
+      return (false, _defaultOf(eltType));
+    }
+
+    if headBlock.isEmpty {
+      halt("Iterating over 1 element with headBlock empty but nElems is ", nElems.read());
+    }
+
+    var elem = headBlock.pop();
+    nElems.sub(1);
+
+    // Fix list if we consumed last one...
+    if headBlock.isEmpty {
+      var tmp = headBlock;
+      headBlock = headBlock.next;
+      delete tmp;
+
+      if headBlock == nil then tailBlock = nil;
+    }
+
+    return (true, elem);
+  }
+
+  inline proc addElements(elt : eltType) {
+    var block = tailBlock;
+
+    // Empty? Create a new one of initial size
+    if block == nil then {
+      tailBlock = new BagSegmentBlock(eltType, INITIAL_BLOCK_SIZE);
+      headBlock = tailBlock;
+      block = tailBlock;
+    }
+
+    // Full? Create a new one double the previous size
+    if block.isFull {
+      block.next = new BagSegmentBlock(eltType, block.cap * 2);
+      tailBlock = block.next;
+      block = block.next;
+    }
+
+    block.push(elt);
+    nElems.add(1);
+  }
+
+  inline proc addElements(elts) {
+    for elt in elts do addElements(elt);
+  }
+}
+
+/*
+  We maintain a multiset 'bag' per node. Each bag keeps a handle to it's parent,
+  which is required for work stealing.
+*/
+pragma "no doc"
+class Bag {
+  type eltType;
+
+  // A handle to our parent 'distributed' bag, which is needed for work stealing.
+  var parentHandle : DistributedBag(eltType);
+
+  /*
+    Helps evenly distribute and balance placement of elements in a best-effort
+    round-robin approach. In the case where we have parallel enqueues or dequeues,
+    they are less likely overlap with each other. Furthermore, it increases our
+    chance to find our 'ideal' segment.
+  */
+  var startIdxEnq : atomic uint;
+  var startIdxDeq : atomic uint;
+
+  /*
+    If a task makes 2 complete passes (1 best-case, 1 average-case) and has not
+    found enough items, then it may attempt to balance the load distribution.
+    Furthermore, if a task is waiting on a load balance, it may piggyback on the
+    result.
+  */
+  var loadBalanceInProgress : atomic bool;
+  var loadBalanceResult : atomic bool;
+
+
+
+  var maxParallelSegmentSpace = {0 .. #here.maxTaskPar};
+  var segments : [maxParallelSegmentSpace] BagSegment(eltType);
+
+  inline proc nextStartIdxEnq {
+    return (startIdxEnq.fetchAdd(1) % here.maxTaskPar : uint) : int;
+  }
+
+  inline proc nextStartIdxDeq {
+    return (startIdxDeq.fetchAdd(1) % here.maxTaskPar : uint) : int;
+  }
+
+  proc Bag(type eltType, parentHandle) {
+    this.parentHandle = parentHandle;
+  }
+
+  proc add(elt : eltType) : bool {
+    var startIdx = nextStartIdxEnq : int;
+    var phase = ADD_BEST_CASE;
+
+    while true {
+      select phase {

```

Comment:
I think it's fine, if you want to write it that way. The compiler literally translates it into some `if` statements.